### PR TITLE
[SPARK-23053][CORE] taskBinarySerialization and task partitions calculate in DagScheduler.submitMissingTasks should keep the same RDD checkpoint status

### PR DESCRIPTION
**Pull request by ivoson**
## What changes were proposed in this pull request?

When we run concurrent jobs using the same rdd which is marked to do checkpoint. If one job has finished running the job, and start the process of RDD.doCheckpoint, while another job is submitted, then submitStage and submitMissingTasks will be called. In [submitMissingTasks](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L961), will serialize taskBinaryBytes and calculate task partitions which are both affected by the status of checkpoint, if the former is calculated before doCheckpoint finished, while the latter is calculated after doCheckpoint finished, when run task, rdd.compute will be called, for some rdds with particular partition type such as [UnionRDD](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rdd/UnionRDD.scala) who will do partition type cast, will get a ClassCastException because the part params is actually a CheckpointRDDPartition.
This error occurs  because rdd.doCheckpoint occurs in the same thread that called sc.runJob, while the task serialization occurs in the DAGSchedulers event loop.

## How was this patch tested?

the exist uts and also add a test case in DAGScheduerSuite to show the exception case.

### CODE REVIEW

**Comment by ivoson** [comment]
@xuanyuanking could review this please?

**Comment by xuanyuanking** [comment]
reopen this...

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 1]
```diff
@@ -96,6 +98,22 @@ class MyRDD(
...
+}
+
+/** Wrapped rdd with WrappedPartition. */
+class WrappedRDD(parent: RDD[Int]) extends RDD[Int](parent) {
+  protected def getPartitions: Array[Partition] = {
+    parent.partitions.map(p => new WrappedPartition(p))
+  }
+
+  def compute(split: Partition, context: TaskContext): Iterator[Int] = {
+    parent.compute(split.asInstanceOf[WrappedPartition].partition, context)
```
Comment (xuanyuanking) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: I think this line is the key point for `WrppedPartition` and `WrappedRDD`, please give comments for explaining your intention.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 2]
```diff
@@ -2399,6 +2417,93 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
+   * Job two is submitted, and submitMissingTasks is called.
+   * In submitMissingTasks, if taskSerialization is called before doCheckpoint is done,
+   * while part calculates from stage.rdd.partitions is called after doCheckpoint is done,
+   * we may get a ClassCastException when execute the task because of some rdd will do
+   * Partition cast.
+   *
+   * With this test case, just want to indicate that we should do taskSerialization and
+   * part calculate in submitMissingTasks with the same rdd checkpoint status.
+   */
+  test("task part misType with checkpoint rdd in concurrent execution scenes") {
```
Comment (xuanyuanking) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: maybe "SPARK-23053: avoid CastException in concurrent execution with checkpoint" better?

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 3]
```diff
@@ -2399,6 +2417,93 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
+        val func = (ctx: TaskContext, it: Iterator[Int]) => cleanedFunc(it)
+        val taskBinaryBytes = JavaUtils.bufferToArray(
+          ser.serialize((rdd, func): AnyRef))
+        semaphore2.release()
+        semaphore1.acquire()
+        // Part calculated with rdd checkpoint already finished.
+        val (taskRdd, taskFunc) = ser.deserialize[(RDD[Int], (TaskContext, Iterator[Int]) => Unit)](
+          ByteBuffer.wrap(taskBinaryBytes), Thread.currentThread.getContextClassLoader)
+        val part = rdd.partitions(0)
+        intercept[ClassCastException] {
```
Comment (xuanyuanking) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: I think this not a "test", this just a "reproduce" for the problem you want to fix. We should prove your code added in `DAGScheduler.scala` can fix that problem and with the original code base, a `ClassCastException` raised.

**Review by xuanyuanking — CHANGES_REQUESTED** [review]

**Comment by xuanyuanking** [comment]
test this please

**Comment by xuanyuanking** [comment]
ok to test

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala (hunk 2 above)
Comment (ivoson) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: thanks for the suggest.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala (hunk 3 above)
Comment (ivoson) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: it is a reproduce case, i will fix this.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala (hunk 1 above)
Comment (ivoson) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: thanks for the comment, i will work on this.

**Review by xuanyuanking — APPROVED** [review]
LGTM, cc @zsxwing @jerryshao @gatorsmile @cloud-fan 
hi Shixiong, Saisai, Xiao and Wenchen. We found this streaming job problem in Baidu practical, Spark version 2.1. This triggered by streaming checkpoint workload and the fix patch mainly on core module. Please give a review. Thanks :)

**Comment by xuanyuanking** [comment]
@ivoson Tengfei, please post the full stack trace of the `ClassCastException`.

**Comment by ivoson** [comment]
This is the stack trace of the Exception.

```
java.lang.ClassCastException: org.apache.spark.rdd.CheckpointRDDPartition cannot be cast to org.apache.spark.streaming.rdd.MapWithStateRDDPartition
at org.apache.spark.streaming.rdd.MapWithStateRDD.compute(MapWithStateRDD.scala:152)
at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:323)
at org.apache.spark.rdd.RDD$$anonfun$8.apply(RDD.scala:336)
at org.apache.spark.rdd.RDD$$anonfun$8.apply(RDD.scala:334)
at org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:957)
at org.apache.spark.storage.BlockManager$$anonfun$doPutIterator$1.apply(BlockManager.scala:948)
at org.apache.spark.storage.BlockManager.doPut(BlockManager.scala:888)
at org.apache.spark.storage.BlockManager.doPutIterator(BlockManager.scala:948)
at org.apache.spark.storage.BlockManager.getOrElseUpdate(BlockManager.scala:694)
at org.apache.spark.rdd.RDD.getOrCompute(RDD.scala:334)
at org.apache.spark.rdd.RDD.iterator(RDD.scala:285)
```

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 4]
```diff
@@ -1016,15 +1016,23 @@ class DAGScheduler(
...
       // For ResultTask, serialize and broadcast (rdd, func).
-      val taskBinaryBytes: Array[Byte] = stage match {
-        case stage: ShuffleMapStage =>
-          JavaUtils.bufferToArray(
-            closureSerializer.serialize((stage.rdd, stage.shuffleDep): AnyRef))
-        case stage: ResultStage =>
-          JavaUtils.bufferToArray(closureSerializer.serialize((stage.rdd, stage.func): AnyRef))
+      var taskBinaryBytes: Array[Byte] = null
+      // Add synchronized block to avoid rdd deserialized from taskBinaryBytes has diff checkpoint
+      // status with the rdd when create ShuffleMapTask or ResultTask.
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: I'd reword this a bit:

taskBinaryBytes and partitions are both effected by the checkpoint status.  We need this synchronization in case another concurrent job is checkpointing this RDD, so we get a consistent view of both variables.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 5]
```diff
@@ -2399,6 +2424,115 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
+   * Partition cast.
+   *
+   * With this test case, just want to indicate that we should do taskSerialization and
+   * part calculate in submitMissingTasks with the same rdd checkpoint status.
+   */
+  test("SPARK-23053: avoid ClassCastException in concurrent execution with checkpoint") {
+    // set checkpointDir.
+    val tempDir = Utils.createTempDir()
+    val checkpointDir = File.createTempFile("temp", "", tempDir)
+    checkpointDir.delete()
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: why do you make a tempfile for the checkpoint dir and then delete it?  why not just  `checkpointDir = new File(tempDir, "checkpointing")`?  Or even just `checkpointDir = Utils.createTempDir()`?

(CheckpointSuite does this so it can call `sc.setCheckpointDir`, but you're not doing that here)

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 6]
```diff
@@ -2399,6 +2424,115 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
+    // Semaphores to control the process sequence for the two threads below.
+    val semaphore1 = new Semaphore(0)
+    val semaphore2 = new Semaphore(0)
+
+    val rdd = new WrappedRDD(sc.makeRDD(1 to 100, 4))
+    rdd.checkpoint()
+
+    val checkpointRunnable = new Runnable {
+      override def run() = {
+        // Simply simulate what RDD.doCheckpoint() do here.
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: I'd remove "simply" here and elsewhere in comments.  Also "do" -> "does"

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 7]
```diff
@@ -2399,6 +2424,115 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
+          }
+        }
+
+        val newRDD = checkpointData.doCheckpoint()
+
+        // Release semaphore1 after job triggered in checkpoint finished, so that taskBinary
+        // serialization can start.
+        semaphore1.release()
+        // Wait until taskBinary serialization finished in submitMissingTasksThread.
+        semaphore2.acquire()
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: this would be a bit easier to follow if you rename your semaphores a bit.

`semaphore1` -> `doCheckpointStarted`
`semaphore2` -> `taskBinaryBytesFinished`

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 8]
```diff
@@ -2399,6 +2424,115 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
+
+        // Update our state and truncate the RDD lineage.
+        RDDCheckpointData.synchronized {
+          checkpointData.cpRDD = Some(newRDD)
+          checkpointData.cpState = CheckpointState.Checkpointed
+          rdd.markCheckpointed()
+        }
+        semaphore1.release()
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: and then this would be another semaphore `checkpointStateUpdated`

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 9]
```diff
@@ -2399,6 +2424,115 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
+        // partition is calculated here.
+        val correctPart = rdd.partitions(0)
+
+        // Release semaphore2 so changing checkpoint status to Checkpointed will be done in
+        // checkpointThread.
+        semaphore2.release()
+        // Wait until checkpoint status changed to Checkpointed in checkpointThread.
+        semaphore1.acquire()
+
+        // Part calculated with rdd checkpoint already finished.
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: I'd add a comment above this:

Now we're done simulating the interleaving that might happen within the scheduler -- we'll check to make sure the final state is OK by simulating a couple steps that normally happen on the executor.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 10]
```diff
@@ -2399,6 +2424,115 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
+        val errPart = rdd.partitions(0)
+
+        // TaskBinary will be deserialized when run task in executor.
+        val (taskRdd, taskFunc) = ser.deserialize[(RDD[Int], (TaskContext, Iterator[Int]) => Unit)](
+          ByteBuffer.wrap(taskBinaryBytes), Thread.currentThread.getContextClassLoader)
+
+        val taskContext = mock(classOf[TaskContext])
+        doNothing().when(taskContext).killTaskIfInterrupted()
+
+        // ClassCastException is expected with errPart.
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: I think this is a bit easier to follow if you say

Make sure our test case is setup correctly -- we expect a ClassCastException here if we use the `rdd.partitions` *after* checkpointing was done, but our binary bytes is from before it finished.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 11]
```diff
@@ -2399,6 +2424,115 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
+        taskRdd.iterator(correctPart, taskContext)
+      }
+    }
+
+    new Thread(checkpointRunnable).start()
+    val submitMissingTasksThread = new Thread(submitMissingTasksRunnable)
+    submitMissingTasksThread.start()
+    submitMissingTasksThread.join()
+
+    Utils.deleteRecursively(tempDir)
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: this should be done in a `finally`.

**Review by squito — CHANGES_REQUESTED** [review]
Thanks a lot for finding & working on this @ivoson.  Overall this fix looks great, I just have some small comments for clarity.

Two other small asks:

1) Both in the PR description and the JIRA, can you add that this error occurs because `rdd.doCheckpoint` occurs in the same thread that called `sc.runJob`, while the task serialization occurs in the DAGSchedulers event loop?

2) I saw you put the stack trace in a PR comment, can you also put it on the JIRA?  I think that is easier to find for most users.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala (hunk 5 above)
Comment (ivoson) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: check the code again and yes ```checkpointDir = Utils.createTempDir()``` is enough for this case, will fix this.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala (hunk 11 above)
Comment (ivoson) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: will fix this.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala (hunk 6 above)
Comment (ivoson) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: will fix this.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala (hunk 8 above)
Comment (ivoson) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: thanks for the advise

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala (hunk 7 above)
Comment (ivoson) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: thanks for the advise. will fix this.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala (hunk 10 above)
Comment (ivoson) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: thanks for the advise, it is really helpful for understanding, will update this.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala (hunk 9 above)
Comment (ivoson) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: thanks for the advise, it is really helpful for understanding, will update this.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 4 above)
Comment (ivoson) [reply core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: thanks for the advise.

**Comment by ivoson** [comment]
@squito Hi Rashid, thanks for your review and advise. The PR description and JIRA have been updated , and also put the stack trace on the JIRA. 
The last commit addresses the comments you left. Thanks again for the advise.

File: core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala [hunk 12]
```diff
@@ -2399,6 +2424,121 @@ class DAGSchedulerSuite extends SparkFunSuite with LocalSparkContext with TimeLi
...
+   * Job two is submitted, and submitMissingTasks is called.
+   * In submitMissingTasks, if taskSerialization is called before doCheckpoint is done,
+   * while part calculates from stage.rdd.partitions is called after doCheckpoint is done,
+   * we may get a ClassCastException when execute the task because of some rdd will do
+   * Partition cast.
+   *
+   * With this test case, just want to indicate that we should do taskSerialization and
+   * part calculate in submitMissingTasks with the same rdd checkpoint status.
+   */
+  test("SPARK-23053: avoid ClassCastException in concurrent execution with checkpoint") {
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: hi @ivoson  -- I'm really sorry but I only just realized that this "test" is really just a repro, and it passes both before and after the actual code changes, since you've replicated the internal logic we're fixing.  As such, I don't think its actually useful as a test case -- perhaps it should get added to the jira as a repro.

I appreciate the work that went into writing this as it helped make the issue clear to me.  I am not sure if there is a good way to test this.  If we can't come up with anything, we should just commit your actual fix, but give me a day or two to think about it ...

Comment (ivoson) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: @squito thanks for reply. I understand this, technically it may not be a UT case, just simulate the scene with exception. I also wonder if there is a good way to test this.

Comment (squito) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: hi @ivoson -- I haven't come up with a better way to test this, so I think for now you should

(1) change the PR to *only* include the changes to the DAGScheduler (also undo the `protected[spark]` changes elsewhere)
(2) put this repro on the jira as its a pretty good for showing whats going on.

if we come up with a way to test it, we can always do that later on.

thanks and sorry for the back and forth

Comment (ivoson) [reply core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala]: hi @squito , it's fine. The pr and jira have been updated. Thanks for your patient and review.

**Review by squito — APPROVED** [review]
lgtm

**Comment by squito** [comment]
merged to master / 2.3 / 2.2

I hit a merge conflict trying to merge to 2.1 -- feel free to open another PR for that version.

**Comment by ivoson** [comment]
thank you for reviewing this @squito
